### PR TITLE
[DeviceSanitizer] Memory overhead statistics

### PR DESCRIPTION
**Pull request by AllanZyne**
UR: https://github.com/oneapi-src/unified-runtime/pull/1869

Statistic memory overhead (usm redzone + shadow memory) on each context, enabled by exporting asan flag UR_LAYER_ASAN_OPTIONS=print_stats:1.

### CODE REVIEW

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// REQUIRES: linux
+// RUN: %{build} %device_asan_flags -O2 -g -o %t
+// RUN: env SYCL_PREFER_UR=1 %{run} %t 2>&1 | FileCheck %s
```
Comment (yingcong-wu) [inline sycl/test-e2e/AddressSanitizer/common/statistics.cpp:3]: `SYCL_PREFER_UR=1` is not needed anymore.

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// RUN: env SYCL_PREFER_UR=1 UR_LAYER_ASAN_OPTIONS=print_stats:1 %{run} %t 2>&1 | FileCheck --check-prefixes CHECK-STATS %s
```
Comment (yingcong-wu) [inline sycl/test-e2e/AddressSanitizer/common/statistics.cpp:4]: if use env, then please use env for each environment variable.

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#include <sycl/usm.hpp>
+
+/// This test is used to check enabling/disabling kernel debug message
+/// We always use "[kernel]" prefix in kernel debug message
```
Comment (yingcong-wu) [inline sycl/test-e2e/AddressSanitizer/common/statistics.cpp:8]: Not related comment?

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp (hunk 2 above)
Comment (AllanZyne) [inline sycl/test-e2e/AddressSanitizer/common/statistics.cpp:4]:
```suggestion
// RUN: %{run} %t 2>&1 | FileCheck %s
// RUN: env UR_LAYER_ASAN_OPTIONS=print_stats:1 %{run} %t 2>&1 | FileCheck --check-prefixes CHECK-STATS %s
```

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// REQUIRES: linux
+// RUN: %{build} %device_asan_flags -O2 -g -o %t
+// RUN: %{run} %t 2>&1 | FileCheck %s
+// RUN: env UR_LAYER_ASAN_OPTIONS=print_stats:1 %{run} %t 2>&1 | FileCheck --check-prefixes CHECK-STATS %s
+#include <sycl/usm.hpp>
+
+/// This test is used to check enabling/disabling kernel debug message
+/// We always use "[kernel]" prefix in kernel debug message
```
Comment (AllanZyne) [inline sycl/test-e2e/AddressSanitizer/common/statistics.cpp:8]:
```suggestion
/// This test is used to check enabling/disabling memory overhead statistics
/// We always use "Stats" prefix in statistics message like asan
```

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp (hunk 2 above)
Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/common/statistics.cpp]: done

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp (hunk 1 above)
Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/common/statistics.cpp]: done

File: sycl/test-e2e/AddressSanitizer/common/statistics.cpp (hunk 3 above)
Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/common/statistics.cpp]: modified

**Review by yingcong-wu — APPROVED** [review]
lgtm

**Comment by AllanZyne** [comment]
Hi @intel/unified-runtime-reviewers, please review

Those CI failures are infrastructure issues.

**Comment by AllanZyne** [comment]
Kindly ping @intel/unified-runtime-reviewers

File: sycl/cmake/modules/FetchUnifiedRuntime.cmake [hunk 5]
```diff
@@ -116,14 +116,14 @@ if(SYCL_UR_USE_FETCH_CONTENT)
...
-  #   fix perf regression in tracing layer
-  set(UNIFIED_RUNTIME_TAG 7384e2d7b908de0b2bca9f3c57827ea84698864e)
+  set(UNIFIED_RUNTIME_REPO "https://github.com/AllanZyne/unified-runtime.git")
+  # commit 7e38af77174f92bf4025122b1fa8e3fc49f3ba29
+  # Merge: ddafd29b 6c7d7a76
+  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
+  # Date:   Mon Jul 15 15:14:47 2024 +0100
+  #     Merge pull request #1863 from kbenzie/benie/main-reverts
+  #     Revert "Merge pull request #1849 from ayylol/cl-subgroupsizes"
+  set(UNIFIED_RUNTIME_TAG review/yang/statistics)
```
Comment (aarongreig) [inline sycl/cmake/modules/FetchUnifiedRuntime.cmake:126]: https://github.com/oneapi-src/unified-runtime/pull/1869 is merged, please update

**Review by kbenzie — APPROVED** [review]

**Comment by kbenzie** [comment]
The UR change was already introduced as part of #15480.

**Comment by kbenzie** [comment]
@intel/llvm-gatekeepers please merge

**Pull request merged**